File: Chapter11/C#/SystemBrowser/Builder.cs
using System;
using System.Windows.Forms;
using System.Threading;

namespace SystemBrowser
{
  /// <summary>
  /// This class builds all the top-level objects of the application
  /// </summary>
  public class Builder
  {
    // there can only be one instance of the Builder
    static Builder singleton;
    static public Builder Singleton
    {
      get 
      {
        if (singleton == null)
          singleton = new Builder();
        return singleton;
      }
    }

    public Binder binder;
    public UserSettings userSettings;

    // UI elements
    public FormSplash formSplash;
    public FormMain formMain;
    public FormMenuToolBar formMenuToolBar;
    public StatusBar statusBar;

    // Workers
    public NavigatorFolders navigatorFolders;
    public NavigatorSearch navigatorSearch;
    public ContentFileList contentFolders;
    public ContentSearchResults contentSearch;

    // Coordinators
    public CoordinatorStartup startup;   // lifecycle: starting up state 
    public CoordinatorShutdown shutdown; // lifecycle: shutting down state
    public CoordinatorCruise cruise;     // lifecycle: cruising state
    public CoordinatorSearch coordinatorSearch;

    public Builder()
    {
      formSplash = new FormSplash();
      binder = new Binder(this);
      startup = new CoordinatorStartup();
      userSettings = new UserSettings();
    }

    public void Build()
    {
      // create the navigators for the left pane
      navigatorFolders = new NavigatorFolders();
      navigatorSearch = new NavigatorSearch();

      // create the content managers for the right pane
      contentFolders = new ContentFileList();
      contentSearch = new ContentSearchResults();

      // UI elements
      formMenuToolBar = new FormMenuToolBar();
      statusBar = new StatusBar();
      formMain = new FormMain();

      // Coordinators
      cruise = new CoordinatorCruise();
      shutdown = new CoordinatorShutdown();
      coordinatorSearch = new CoordinatorSearch(formMain);

      // move the navigators and content viewers into FormMain
      formMain.NavigatorFolders = navigatorFolders;
      formMain.NavigatorSearch = navigatorSearch;
      formMain.ContentFolders = contentFolders;
      formMain.ContentSearch = contentSearch;

      // move the menu, toolbar and statusbar into FormMain
      formMain.Menu = formMenuToolBar.mainMenu;
      formMain.Toolbar = formMenuToolBar.panelToolBar;
      formMain.Statusbar = statusBar;
    }
  }
}

File: Chapter11/VB/SystemBrowser/Builder.vb
Imports System.Threading

' This class builds all the top-level objects of the application

Public Class Builder
  ' there can only be one instance of the Builder
  Private Shared _singleton As Builder
  Public Shared ReadOnly Property Singleton() As Builder
    Get
      If _singleton Is Nothing Then
        _singleton = New Builder
      End If
      Return _singleton
    End Get
  End Property

  Public _binder As Binder
  Public _userSettings As UserSettings

  ' UI elements
  Public _formSplash As FormSplash
  Public _formMain As FormMain
  Public _formMenuToolBar As FormMenuToolBar
  Public _statusBar As StatusBar

  ' Workers
  Public _navigatorFolders As NavigatorFolders
  Public _navigatorSearch As NavigatorSearch
  Public _contentFolders As ContentFileList
  Public _contentSearch As ContentSearchResults

  ' Coordinators
  Public _startup As CoordinatorStartup 'lifecycle: starting up state
  Public _shutdown As CoordinatorShutdown 'lifecycle: shutting down state
  Public _cruise As CoordinatorCruise 'lifecycle: cruising state
  Public _coordinatorSearch As CoordinatorSearch '

  Public Sub New()
    _formSplash = New FormSplash
    _binder = New Binder(Me)
    _startup = New CoordinatorStartup
    _userSettings = New UserSettings
  End Sub

  Public Sub Build()
    ' create the navigators for the left pane
    _navigatorFolders = New NavigatorFolders
    _navigatorSearch = New NavigatorSearch

    ' create the content managers for the right pane
    _contentFolders = New ContentFileList
    _contentSearch = New ContentSearchResults

    ' UI elements
    _formMenuToolBar = New FormMenuToolBar
    _statusBar = New StatusBar
    _formMain = New FormMain

    ' Coordinators
    _cruise = New CoordinatorCruise
    _shutdown = New CoordinatorShutdown
    _coordinatorSearch = New CoordinatorSearch(_formMain)

    ' move the navigators and content viewers into FormMain
    _formMain.NavigatorFolders = _navigatorFolders
    _formMain.NavigatorSearch = _navigatorSearch
    _formMain.ContentFolders = _contentFolders
    _formMain.ContentSearch = _contentSearch

    ' move the menu, toolbar and statusbar into FormMain
    _formMain.Menu = _formMenuToolBar.mainMenu
    _formMain.Toolbar = _formMenuToolBar.panelToolBar
    _formMain.Statusbar = _statusBar
  End Sub
End Class
